### ISSUE TITLE
Please bring back the `helm upgrade --recreate-pods` flag

### ISSUE THREAD

**Issue by jpatokal** [helm/helm#8271]
The [Helm tips and tricks page](https://helm.sh/docs/howto/charts_tips_and_tricks/) helpfully notes that configuration changes without a deployment change can easily result in an inconsistent deployment:

> Often times ConfigMaps or Secrets are injected as configuration files in containers or there are other external dependencies changes that require rolling pods. Depending on the application a restart may be required should those be updated with a subsequent helm upgrade, but if the deployment spec itself didn't change the application keeps running with the old configuration resulting in an inconsistent deployment.

This is a real issue, and in Helm 2, the `--recreate-pods` option let you easily recreate the pods, neatly solving this.  But in Helm 3, this has been removed in favor of telling people to update their Deployments with a random annotation.  I understand the theoretical purity of making this declarative, but in practice, when you're debugging a values YAML, this is often quite inconvenient:

1. If I'm installing directly from a 3rd-party server, I now need to go fork the chart, make a local copy, edit the deployment and change the command to point to it, all just to add that annotation.
2. I don't necessarily *always* want to recreate pods: in fact, if I'm tweaking the values until they work, odds are I only want to do it this once.

So pretty please, un-deprecate the `--recreate-pods` flag.

**Comment by bacongobbler** [comment]
For context, the --recreate-pods flag was deprecated for good reason:

- the use cases are all handled by cleaner methods discovered by the community (like the annotation)
- the code inside --recreate-pods was flaky in that it deleted any pods that matched the labels of the workload, even if that pod was manually instantiated with a `kubectl run`. This could cause irrecoverable state loss
- `--recreate-pods` was an all-or-nothing - you either recreate all the pods in the chart, or none at all.
- `--recreate-pods` did not work with other feature flags well, like `--wait` and the annotation described above (with both in place, I believe it would cause a pod to be re-deployed twice)

@thomastaylor312 might be able to give a larger explanation on the reasons for the removal. My understanding for the removal was that all use cases currently handled by `--recreate-pods` can be handled better through other means.

For reference: https://github.com/helm/helm/pull/6463

All that being said, I'm not really in favour of re-introducing a flag that was deemed unmaintainable. However, if there are use cases that are currently not being handled by existing means, let's talk about them. From your feedback @jpatokal it seems more like you're talking about developing a chart locally. Is that correct?

**Comment by mattfarina** [comment]
@jpatokal Have you tried using `kubectl` to delete the posts you need? Whether we get to it or not, any change would take some time. Is `kubectl` an option to use in the short term, at the least?

**Comment by thomastaylor312** [comment]
@bacongobbler is correct here:

> My understanding for the removal was that all use cases currently handled by --recreate-pods can be handled better through other means.

If you need a simple annotation add, you could also take advantage of the `--post-renderer` flag to pass it to an external process (like customize) to drop in an annotation as well. The `kubectl` suggestion is also a viable workaround

**Comment by jpatokal** [comment]
@mattfarina Yes, I'm currently using `kubectl delete pods -n foobar --all` (since the chart happens to install into its own namespace), but this is inelegant to say the least.  

My concrete usecase is simply deploying [Splunk Connect for Kubernetes](https://github.com/splunk/splunk-connect-for-kubernetes/tree/develop/helm-chart/splunk-connect-for-kubernetes), where the charts come as a binary blob (tgz) and you just customize the values YAML for your Splunk deployment:

`helm install my-splunk-connect -f my_values.yaml https://github.com/splunk/splunk-connect-for-kubernetes/releases/download/1.4.1/splunk-connect-for-kubernetes-1.4.1.tgz
`
So for avoidance of doubt, I do not want or need to touch the charts themselves.    But every time I tweak the values and update, I also need to recreate the pods to get them to read the new resulting config.

**Comment by bacongobbler** [comment]
I'm confused. All of the workloads defined within the charts under that repository all have a checksum annotation. Any time configuration changes, the pods should be recreated. If configuration does not change, the pods will not be recreated.

https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-logging/templates/daemonset.yaml#L24

https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-metrics/templates/daemonset.yaml#L29

https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-objects/templates/deployment.yaml#L26

What is the issue you are trying to solve?

**Comment by jpatokal** [comment]
I have just now edited values.yaml and run helm upgrade, and as you can see, the pods did *not* recreate:

```
$ kubectl -n splunk get pods
NAME                                                           READY   STATUS    RESTARTS   AGE
my-splunk-splunk-kubernetes-logging-9bxkm                  1/1     Running   0          45h
[...]
$ vi values.yaml
$ helm upgrade my-splunk -f values.yaml --namespace splunk https://github.com/splunk/splunk-connect-for-kubernetes/releases/download/1.4.1/splunk-connect-for-kubernetes-1.4.1.tgz
Release "my-splunk" has been upgraded. Happy Helming!
$ kubectl -n splunk get pods
NAME                                                           READY   STATUS    RESTARTS   AGE
my-splunk-splunk-kubernetes-logging-9bxkm                  1/1     Running   0          45h
```

**Comment by lavleenn** [comment]
checksum annotation is definitely an acceptable solution when we edit configmap and secrets. Could any of you please provide your inputs on how would we get the pods restarted using a single helm command in a scenario where we edit the environment variables and want our pods to be restarted for them to be updated in the application. 

Or Can any of please suggest a better approach to pass environment variables to the application.

**Comment by Woile** [comment]
I've done the same and also my deployment nor pod are updated.
I've changed `appVersion` and `version` in the `Chart.yaml` and kubernetes still has the old version, I have to remove them manually:

```bash
helm upgrade --namespace=foobar --install foobar charts/foobar
```

**Comment by swarthofole** [comment]
@Woile and @m1pl I followed the official guidelines and it's working for me. have you sort it out? I am curious about any bugs if there are some...

**Comment by cabrinoob** [comment]
The problem with **--recreate-pods** flag in my opinion is that it completely bypass the rolling update strategy of your deployment. It kills all pods and wait that the replicaset controller does its job to raise new ones.

For development purpose this flag is usefull, but in production environment it causes downtime errors.

Personaly, if have I have an update that does not trigger a restart of my pods, I prefer do it in 2 times : 

    helm upgrade myStack ./myStackChart --set ...
    kubectl rollout restart deploy myStack-deploy

This way my services are still highly availables.

**Comment by foogunlana** [comment]
Same problem is similar. I've installed the `https://charts.gitlab.io/gitlab-runner` chart as a dependency to mine.
```
dependencies:
  - name: gitlab-runner
    version: 0.22.0
    repository: https://charts.gitlab.io
```
I've then changed the values passed in from:
```
gitlab-runner:
  gitlabUrl: https://gitlab.com/
  concurrent: 10
```
to 
```
gitlab-runner:
  gitlabUrl: https://gitlab.com/
  concurrent: 10
  cache:
     ...
```

I can clearly see a diff (using `helm diff`) when upgrading. But the deployment doesn't change. It's likely because the gitlab runner helm chart does not roll the deployment itself (or allow me to do so): https://gitlab.com/gitlab-org/charts/gitlab-runner/-/blob/master/templates/deployment.yaml. Seems like I have to delete and recreate it.

Is there any way to get it to role without using a post processor?

**Comment by andyliddle** [comment]
Is there any way to do this for a dependency chart?  I'm trying to ensure that when a release is updated it restarts all of the pod including all child charts.

for example,  PostgreSQL,  it would be good to redeploy the DB container force a init scripts to rerun and get a fresh DB.  I can set the value of the annotation in the values but it's not dynamic.

trying to avoid help upgrade --set as its deployed automatically (Octopus Deploy)

**Comment by imjuanleonard** [comment]
We plan to upgrade some of our pods using helm, although the helm said its already deployed, but the pod didn't get to restart. This causes confusion as the state drifted from the configuration and real one running.

Can we return back the helm recreate pods?

**Comment by omerozery** [comment]
Random annotations and/or files checksum simply don't provide the same functionality as 'kubectl rollout restart'
if you don't want to bring back the --recreate-pods because it causes downtime then give some helm shell flag for 'kubectl rollout restart' otherwise people will just turn to kubectl command for release lifecycle related functionalities where helm should play the part.

**Comment by kayrus** [comment]
I experience the same issue. Good example is nginx ingress controller and custom template, which is loaded only during the first start and nginx deployment doesn't have an annotation with a template hash. https://github.com/kubernetes/ingress-nginx/issues/501#issuecomment-780724391

**Comment by dcserrot** [comment]
Which alternative do we have to refresh docker images?

**Comment by mih-kopylov** [comment]
Does ’kubectl rollout restart’ help?

**Comment by enoliveira** [comment]
Same issue. Pods are not recreating. How can I get the pods to recreate after running helm?

**Comment by bacongobbler** [comment]
To answer the original question the OP raised... We've had several core maintainers (and a few community members) chime in and say we're not interested in re-introducing the `--recreate-pods` flag. Therefore I am going ahead and closing this as we do not plan on reintroducing it for the reasons stated above.

If you are having issues related to pods not restarting after an update and you followed the documentation, please open a new issue and we can follow up individually to see if there are scenarios where an update may not occur and we need to document that.

@enoliveira can you please open a new issue describing the steps you followed along with your use case so we can better understand your situation? That would be most helpful.

Thanks everyone.

**Comment by CharlieReitzel** [comment]
`helm upgrade` simply doesn't work with the `Deployment.spec.strategy = Recreate`.

My specific use case is when there is a database schema change.   I have a complex container I am tasked with getting running under Kubernetes (check) and managed using Helm (not check, not yet).

For basic install, we're good.   Using `Recreate`, I was able to get a database schema update to work, so long as the changes were applied "in place" (i.e. in the same PV/PVC).   Now, we have been asked to make a backup of the data before applying schema changes.

This means mounting the "old" PV/PVC as the "previous version" and a new PV/PVC as the "current version" and copying the data from the old to the new in a pre-upgrade hook.

Now there are all kinds of problems.   Basically, Helm does not appear to have any meaningful order in which it applies the patches it generates.  So, rather than starting with the deployment (which has a new image version), it starts with the PVCs (volume names have changed per above).  But you can't change PVCs that are still in use.  The PVCs are still in use because the old pods are still running.

Long story short: `--recreate-pods` is necessary to work with `Deployment.spec.strategy = Recreate`.

**Comment by CharlieReitzel** [comment]
The alternative would be to give Chart developers a way to control the order patches are applied.  If I could tell Helm "start with the Deployment", that would cause the pods to be stopped (strategy = Recreate) and, subsequently, the PVCs would be patchable.    

That is perhaps even _less_ declarative than `--recreate-pods`.  But most declarative systems end up adding some imperative syntax to support edge (and not-so-edge) cases.  SQL famously so.

**Comment by shanthkumar079** [comment]
We are having a stateful set with a rolling update strategy, also the rollme and timestamp in the template annotations,  The pods are not getting re-created without the re-create-pods flag.

**Comment by aman0303** [comment]
Adding annotation in pod's spec metadata worked for me

**Comment by dambrogia** [comment]
This has worked wonderfully for me:

Using the `Release.Revision` value of the [built-in `Release` object](https://helm.sh/docs/chart_template_guide/builtin_objects/) worked great for me.



```
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: "{{ .Values.namespace }}"
  name: my-custom-deployment
  labels:
    deployment: my-custom-deployment
spec:
  replicas: {{ .Values.replicaCount }}
  selector:
    matchLabels:
      deployment: my-custom-deployment
  template:
    metadata:
      annotations:
      labels:
        deployment: my-custom-deployment
        deploymentRevisionId: "{{ .Release.Revision }}" # this here updates and forces refresh
    spec:
      restartPolicy: Always
      containers:
      - name: my-custom-deployment
      ...
      ...
```
